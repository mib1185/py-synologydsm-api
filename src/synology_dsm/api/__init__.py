"""Synology API models."""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from synology_dsm import SynologyDSM


class SynoBaseApi:
    """Base api class."""

    def __init__(self, dsm: "SynologyDSM") -> None:
        """Constructor method."""
        self._dsm = dsm
        self._data = {}
