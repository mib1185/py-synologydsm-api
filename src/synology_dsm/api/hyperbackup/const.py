"""Synology HyperBackup API constants."""
PROP_NAME = 'name'
PROP_TRANSFER_TYPE = 'transfer_type'
PROP_LAST_RESULT = 'last_bkp_result'  # Raw previous result
PROP_LAST_BACKUP_TIME = 'last_bkp_time'
PROP_LAST_BACKUP_END_TIME = 'last_bkp_end_time'
PROP_LAST_BACKUP_ERROR = 'last_bkp_error'
PROP_NEXT_BACKUP_TIME = 'next_bkp_time'
PROP_LAST_BACKUP_PROGRESS = 'last_bkp_progress'
PROP_STATE = 'state'
PROP_STATUS = 'status'  # 'raw status' from API.
PROP_TARGET_ID = 'target_id'
PROP_TASKID = 'task_id'
PROP_ONLINE = 'is_online'
PROP_USED_SIZE = 'used_size'
PROP_LAST_PROGRESS = 'last_bkp_progress'

# Hyper backup Task raw status values
PROP_STATUS_NONE = 'none'
PROP_STATUS_WAITING = 'waiting'
PROP_STATUS_BACKUP = 'backup'
PROP_STATUS_DETECT = 'detect'
PROP_STATUS_VER_DEL = 'version_deleting'
PROP_STATUS_PREP_VER_DEL = 'preparing_version_delete'

# Hyper backup raw result values
RESULT_NONE = 'none'
RESULT_DONE = 'done'
RESULT_RESUME = 'resuming'
RESULT_SUSPEND = 'suspend'

# Raw state values
STATE_BACKUP = 'backupable'
STATE_EXPORT = 'exportable'
STATE_IMPORT = 'importable'
STATE_RELINK = 'relinkable'
STATE_ERROR = 'error_detect'
STATE_BROKEN = 'broken'
STATE_RESTORE_ONLY = 'restore_only'
STATE_UNAUTH = 'unauth'
STATE_END_SERVICE = 'endofservice'

# Hyper backup *derived* Health values
HEALTH_GOOD = 'Good'
HEALTH_WARN = 'Warning'
HEALTH_CRIT = 'Error'

# Hyper backup *derived* status values
STATUS_OK = 'OK'
STATUS_RUNNING = 'Backing up'
STATUS_WAITING = 'Waiting to backup'
STATUS_RESUMING = 'Resuming'
STATUS_SUSPENDED = 'Suspended'
STATUS_NEVER_RUN = 'Never backed up'
STATUS_NO_SCHEDULE = 'No schedule set'  # Backup task has run successfully, but there is no active schedule
STATUS_RESTORE_ONLY = 'Restore Only'  # Not backup-able, but can restore
STATUS_UNKNOWN = 'Unhandled/Unknown'
STATUS_ERROR = 'Error'  # An error occurred, or unknown combination of raw status/prev result.


