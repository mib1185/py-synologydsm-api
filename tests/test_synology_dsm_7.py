"""Synology DSM tests."""
import pytest

from synology_dsm.const import API_AUTH
from synology_dsm.exceptions import SynologyDSMLogin2SARequiredException

from . import (
    VALID_HOST,
    VALID_HTTPS,
    VALID_OTP,
    VALID_PASSWORD,
    VALID_PORT,
    VALID_USER_2SA,
    SynologyDSMMock,
)
from .const import DEVICE_TOKEN, SESSION_ID, SYNO_TOKEN


class TestSynologyDSM7:
    """SynologyDSM 7 test cases."""

    @pytest.mark.asyncio
    async def test_login(self, dsm_7):
        """Test login."""
        assert await dsm_7.login()
        assert dsm_7.apis.get(API_AUTH)
        assert dsm_7._session_id == SESSION_ID
        assert dsm_7._syno_token == SYNO_TOKEN

    @pytest.mark.asyncio
    async def test_login_2sa(self):
        """Test login with 2SA."""
        dsm_7 = SynologyDSMMock(
            None,
            VALID_HOST,
            VALID_PORT,
            VALID_USER_2SA,
            VALID_PASSWORD,
            VALID_HTTPS,
        )
        dsm_7.dsm_version = 7
        with pytest.raises(SynologyDSMLogin2SARequiredException) as error:
            await dsm_7.login()
        error_value = error.value.args[0]
        assert error_value["api"] == "SYNO.API.Auth"
        assert error_value["code"] == 403
        assert error_value["reason"] == "One time password not specified"
        assert (
            error_value["details"]
            == "Two-step authentication required for account: valid_user_2sa"
        )

        assert await dsm_7.login(VALID_OTP)

        assert dsm_7._session_id == SESSION_ID
        assert dsm_7._syno_token == SYNO_TOKEN
        assert dsm_7._device_token == DEVICE_TOKEN
        assert dsm_7.device_token == DEVICE_TOKEN

    @pytest.mark.asyncio
    async def test_login_2sa_new_session(self):
        """Test login with 2SA and a new session with granted device."""
        dsm_7 = SynologyDSMMock(
            None,
            VALID_HOST,
            VALID_PORT,
            VALID_USER_2SA,
            VALID_PASSWORD,
            VALID_HTTPS,
            device_token=DEVICE_TOKEN,
        )
        dsm_7.dsm_version = 7
        assert await dsm_7.login()

        assert dsm_7._session_id == SESSION_ID
        assert dsm_7._syno_token == SYNO_TOKEN
        assert dsm_7._device_token == DEVICE_TOKEN
        assert dsm_7.device_token == DEVICE_TOKEN

    @pytest.mark.asyncio
    async def test_upgrade(self, dsm_7):
        """Test upgrade."""
        assert dsm_7.upgrade
        await dsm_7.upgrade.update()
        assert dsm_7.upgrade.update_available
        assert dsm_7.upgrade.available_version == "7.0.1-42218 Update 3"
        assert dsm_7.upgrade.reboot_needed == "now"
        assert dsm_7.upgrade.service_restarts == "some"
        assert dsm_7.upgrade.available_version_details == {
            "buildnumber": 42218,
            "major": 7,
            "micro": 1,
            "minor": 0,
            "nano": 3,
            "os_name": "DSM",
        }

    @pytest.mark.asyncio
    async def test_photos(self, dsm_7):
        """Test photos."""
        assert dsm_7.photos
        albums = await dsm_7.photos.get_albums()

        assert albums
        assert len(albums) == 2
        assert albums[0].album_id == 4
        assert albums[0].name == "Album1"
        assert albums[0].item_count == 3
        assert albums[1].album_id == 1
        assert albums[1].name == "Album2"
        assert albums[1].item_count == 1

        items = await dsm_7.photos.get_items_from_album(albums[0])
        assert items
        assert len(items) == 3
        assert items[0].file_name == "20221115_185642.jpg"
        assert items[0].thumbnail_cache_key == "29807_1668560967"
        assert items[0].thumbnail_size == "xl"
        assert items[1].file_name == "20221115_185643.jpg"
        assert items[1].thumbnail_cache_key == "29808_1668560967"
        assert items[1].thumbnail_size == "m"
        assert items[2].file_name == "20221115_185644.jpg"
        assert items[2].thumbnail_cache_key == "29809_1668560967"
        assert items[2].thumbnail_size == "sm"

        thumb_url = await dsm_7.photos.get_item_thumbnail_url(items[0])
        assert thumb_url
        assert thumb_url == (
            "https://nas.mywebsite.me:443/webapi/entry.cgi?"
            "id=29807&cache_key=29807_1668560967&size=xl&type=unit"
            "&api=SYNO.Foto.Thumbnail&version=2&method=get"
            "&_sid=session_id&SynoToken=Sy%C3%B10_T0k%E2%82%AC%C3%B1"
        )

        items = await dsm_7.photos.get_items_from_search(albums[0])
        assert items
        assert len(items) == 2
        assert items[0].file_name == "search_1.jpg"
        assert items[0].thumbnail_cache_key == "12340_1668560967"
        assert items[1].file_name == "search_2.jpg"
        assert items[1].thumbnail_cache_key == "12341_1668560967"
