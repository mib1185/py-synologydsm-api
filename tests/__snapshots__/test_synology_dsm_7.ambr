# serializer version: 1
# name: TestSynologyDSM7.test_external_usb
  dict({
    'sdq': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='sdq',
      device_manufacturer='Genesys Logic, Inc.',
      device_name='SD Card 1',
      device_partitions=dict({
        'sdq1': SynoUSBStoragePartition(
          filesystem='FAT32',
          fstype='vfat',
          is_mounted=True,
          is_supported=True,
          name_id='sdq1',
          partition_percentage_used=4.0,
          partition_title='SD Card 1 Partition 1',
          share_name='SDKarte',
          status='normal',
        ),
      }),
      device_product_name=None,
      device_progress='',
      device_status='normal',
      device_type='sdCard',
      partitions_all_percentage_used=4.0,
    ),
    'sds': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='sds',
      device_manufacturer='Iomega Corp.',
      device_name='USB Disk 1',
      device_partitions=dict({
        'sds1': SynoUSBStoragePartition(
          filesystem='ntfs',
          fstype='ntfs',
          is_mounted=True,
          is_supported=True,
          name_id='sds1',
          partition_percentage_used=90.4,
          partition_title='USB Disk 1 Partition 1',
          share_name='usbshare1',
          status='usbbackup',
        ),
      }),
      device_product_name='Prestige Portable Hard Drive',
      device_progress='',
      device_status='usbbackup',
      device_type='usbDisk',
      partitions_all_percentage_used=90.4,
    ),
    'usb1': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb1',
      device_manufacturer='PNY',
      device_name='USB Disk 1',
      device_partitions=dict({
        'usb1p1': SynoUSBStoragePartition(
          filesystem='ntfs',
          fstype='ntfs',
          is_mounted=True,
          is_supported=True,
          name_id='usb1p1',
          partition_percentage_used=1.0,
          partition_title='USB Disk 1 Partition 1',
          share_name='usbshare1-1',
          status='normal',
        ),
        'usb1p2': SynoUSBStoragePartition(
          filesystem='FAT32',
          fstype='vfat',
          is_mounted=True,
          is_supported=True,
          name_id='usb1p2',
          partition_percentage_used=0.0,
          partition_title='USB Disk 1 Partition 2',
          share_name='usbshare1-2',
          status='normal',
        ),
      }),
      device_product_name='Flash Drive',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=0.5,
    ),
    'usb2': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb2',
      device_manufacturer='Western Digital Technologies, Inc.',
      device_name='USB Disk 2',
      device_partitions=dict({
        'usb2p1': SynoUSBStoragePartition(
          filesystem='ext4',
          fstype='ext4',
          is_mounted=True,
          is_supported=True,
          name_id='usb2p1',
          partition_percentage_used=72.0,
          partition_title='USB Disk 2 Partition 1',
          share_name='usbshare2',
          status='normal',
        ),
      }),
      device_product_name='easystore 264D',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=72.0,
    ),
    'usb3': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb3',
      device_manufacturer='SanDisk Corp.',
      device_name='USB Disk 3',
      device_partitions=dict({
        'usb3p1': SynoUSBStoragePartition(
          filesystem='FAT32',
          fstype='vfat',
          is_mounted=True,
          is_supported=True,
          name_id='usb3p1',
          partition_percentage_used=5.1,
          partition_title='USB Disk 3 Partition 1',
          share_name='usbshare3-1',
          status='normal',
        ),
        'usb3p2': SynoUSBStoragePartition(
          filesystem='hfsplus',
          fstype='hfsplus',
          is_mounted=True,
          is_supported=True,
          name_id='usb3p2',
          partition_percentage_used=11.9,
          partition_title='USB Disk 3 Partition 2',
          share_name='usbshare3-2',
          status='normal',
        ),
      }),
      device_product_name=' SanDisk 3.2Gen1',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=11.9,
    ),
    'usb4': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb4',
      device_manufacturer='PNY',
      device_name='USB Disk 4',
      device_partitions=dict({
        'usb4p1': SynoUSBStoragePartition(
          filesystem='exfat',
          fstype='exfat',
          is_mounted=True,
          is_supported=True,
          name_id='usb4p1',
          partition_percentage_used=0.0,
          partition_title='USB Disk 4 Partition 1',
          share_name='usbshare4',
          status='normal',
        ),
      }),
      device_product_name='USB 3.0 FD',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=0.0,
    ),
    'usb5': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb5',
      device_manufacturer='SanDisk Corp.',
      device_name='USB Disk 5',
      device_partitions=dict({
        'usb5p1': SynoUSBStoragePartition(
          filesystem='FAT32',
          fstype='vfat',
          is_mounted=True,
          is_supported=True,
          name_id='usb5p1',
          partition_percentage_used=0.0,
          partition_title='USB Disk 5 Partition 1',
          share_name='usbshare5-1',
          status='normal',
        ),
        'usb5p2': SynoUSBStoragePartition(
          filesystem='FAT32',
          fstype='vfat',
          is_mounted=True,
          is_supported=True,
          name_id='usb5p2',
          partition_percentage_used=84.4,
          partition_title='USB Disk 5 Partition 2',
          share_name='usbshare5-2',
          status='normal',
        ),
      }),
      device_product_name='Cruzer Titanium+',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=0.7,
    ),
    'usb6': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb6',
      device_manufacturer='Western Digital Technologies, Inc.',
      device_name='USB Disk 6',
      device_partitions=dict({
        'usb6p1': SynoUSBStoragePartition(
          filesystem='ntfs',
          fstype='ntfs',
          is_mounted=True,
          is_supported=True,
          name_id='usb6p1',
          partition_percentage_used=28.5,
          partition_title='USB Disk 6 Partition 1',
          share_name='usbshare6',
          status='normal',
        ),
      }),
      device_product_name='My Passport Essential (WDBACY)',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=28.5,
    ),
    'usb8': SynoCoreExternalUSBDevice(
      device_formatable=True,
      device_id='usb8',
      device_manufacturer='Genesys Logic, Inc.',
      device_name='USB Disk 8',
      device_partitions=dict({
        'usb8': SynoUSBStoragePartition(
          filesystem='',
          fstype='unknown',
          is_mounted=False,
          is_supported=False,
          name_id='usb8',
          partition_percentage_used=None,
          partition_title='USB Disk 8',
          share_name='',
          status='normal',
        ),
      }),
      device_product_name='All-in-One Cardreader',
      device_progress='',
      device_status='normal',
      device_type='usbDisk',
      partitions_all_percentage_used=None,
    ),
  })
# ---
